### ISSUE TITLE
List `typescript` as `devDependency`

### ISSUE THREAD

**Issue by fenduru** [dsherret/ts-simple-ast#404]
**Is your feature request related to a problem? Please describe.**

I'm trying to create build tooling for other teams at my company, however I would like each project to be in control of which version of typescript is used so we do not have to coordinate every project upgrading their typescript version at the same time.

**Describe the solution you'd like**

The simplest solution would be to remove the `dependency` on typescript so that `require('typescript')` will find the consuming code's copy. The compile API is technically unstable, but this is the approach tools like `ts-loader` take (https://github.com/TypeStrong/ts-loader/blob/master/package.json#L81) and they have been rather successful.

**Describe alternatives you've considered**

Alternative would be to support the consumer passing in an already-imported copy of typescript. I could then do my own peerDependency/require and pass that along.

**Comment by dsherret** [comment]
This library used to support a broad range of typescript compilers, but it turned out to be a nightmare. Libraries like `ts-loader` and others use a very small stable subset of the compiler API so it's easy for them to do that. This library, however, depends on a very large portion of the API and the specific functionality of the compiler its depending on (when upgrading typescript compiler versions, I sometimes have to spend a lot of time fixing new failing unit tests because some functionality changed).

Basically, this wrapper is very intertwined with the specific typescript compiler version.

You can read the full reasoning for why this isn't supported here: https://github.com/dsherret/ts-simple-ast/issues/333#issuecomment-391182952 (took me a while to find this comment).

**Comment by fenduru** [comment]
Gotcha. Honestly I'm really only interested in the easier-to-use compiler API and was hoping to avoid needing to roll it myself. Looks like I'll be giving that a try :)

**Comment by dsherret** [comment]
Yeah, you will get better performance for builds out of just using the compiler API, but it may be very complicated to do certain tasks. If you're only doing AST transformations then using the compiler API is not too bad.

By the way, in most scenarios you'll be able to get away with using this library to analyze/manipulate code written for a lower version of the compiler. There's not many breaking changes that happen in the language. If you're using this library to compile and type check the files then it won't work so nicely though.

**Comment by fenduru** [comment]
Yeah, I'm really just trying to do typechecking/emitting without any sort of AST modifications. I have the language service emitting and whatnot, but as soon as the script I'm compiling uses something from node modules, it fails to find the types from package.json. I know this is not really a ts-simple-ast issue, but was wondering if you happen to know how to get package.json typings working when dealing with the compiler api

**Comment by dsherret** [comment]
Is that a problem when using ts-simple-ast or you're having when using the compiler API? If using the compiler API, there's probably something slightly wrong in the setup. I believe the compiler host passed in when creating the program will be called to search for the module.

Maybe try to copy the code here, but try without implementing the custom module resolution: https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#customizing-module-resolution

Closing this issue for now, but let me know if you have any issues with ts-simple-ast.
